### PR TITLE
ProffieOS 5.x on Teensy4.0

### PR DESCRIPTION
**Pull request by artandtechvof**
Hi Fredrik,

On TRA you offered your help in getting ProffieOS working on Teensy4.0. I managed to get the OctoWS2811 lib for Teensy4.0 working in ProffieOS without any modification to the lib itself, It took some effort to modify the MonopodWSPinBase class to be a sort of wrapper for the OctoWS2811 lib.

I think you might cringe over some of the ways I wrote the code, and use globals to be used in the MonopodWSPinBase class.

Your help would be appreciated in reviewing this code, and help/give hints in getting rid of the 'cringy' parts  ;-)

Steps that I took to get it working for ProffieOS5.x on a Teensy4.0

Make hardware define config
   aat_teensy40_hardware_config.h

Make style define config
  aat_teensy40_config.h"

ProffieOS.ino

  #define CONFIG_FILE "config/aat_teensy40_config.h" //Teensy 4.0

  #ifdef USE_TEENSY4
    #include <Wire.h>  -> native I2C lib is good enough, might be imporved to use DMA, but no need yet.
  #else
    #include <kinetis.h>
  #endif

  addition of Serial4, vs Serial3, beccause Serial3 overlaps with I2S output on T4.0

  #include for MPU6050async

  setup()
    begin() for stock OctoWS2811_imxrt lib

dac.h
  explicit declaration of audioclock for teensy40

i2cbus
  #ifndef USE_TEENSY4
    Wire.setDefaultTimeout(I2C_TIMEOUT_MILLIS * 1000);
  #endif

MPU6050 -> MPU6050async
  rewrote to use ASYNC I2C and interrupt pin, otherwise would time-out after first retraction of the blade

new monopods -> monopods_imxrt.h
  it has become a wrapper only for the stock OctoWS2811 lib

ws2811_blade.h
  #define fot Teensy4.0 to use monopods_imxrt.h

After these modifications, saber will work with ProffieSO and in hilt Cree RGB led + accent leds.

Current progress is still WIP, and is juts proof of concept

ToDo;

- Init OctoWS2811 lib and integration into MonopodWSPinBase class (not calling globals directly from wihtin the class)
- better auto-configuration based on config.h file for number of blade pins and pinlist
- general cleanup of code
- RGBw support (not yet fully integrated into OctoWS2811 for Teensy4.0

	modified:   ProffieOS.ino
	modified:   blades/monopodws.h
	new file:   blades/monopodws_imxrt.h
	modified:   blades/ws2811_blade.h
	modified:   common/i2cbus.h
	new file:   config/aat_teensy40_config.h
	new file:   config/aat_teensy40_hardware_config.h
	modified:   config/default_proffieboard_config.h
	new file:   motion/mpu6050async.h
	modified:   sound/dac.h
	modified:   styles/ignition_delay.h
	modified:   styles/retraction_delay.h

### CODE REVIEW

File: blades/monopodws_imxrt.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  bool IsReadyForBeginFrame() {
+        if (is_first_blade_) {
+            return !leds.busy();//sync to 1st blade
+        }else{
+            return true; // other blades can start filling displayMemory
+        }           
+  }
+  
+  bool IsReadyForEndFrame() {
+        return millis() > (frame_sync_time_ + min_update_time_);//4ms =250fps, 5ms=200fps,
```
Comment (profezzorn) [inline blades/monopodws_imxrt.h:81]: it would probably be better to do this in micros.

File: ProffieOS.ino [hunk 2]
```diff
@@ -1328,6 +1332,18 @@ public:
   static const char* response_footer() { return "-+=END_OUTPUT=+-\n"; }
 };
 
+#ifdef USE_TEENSY4
+class Serial4Adapter {
+public:
+  static void begin() { Serial4.begin(115200); }
+  static bool Connected() { return true; }
+  static bool AlwaysConnected() { return true; }
+  static Stream& stream() { return Serial3; }
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply ProffieOS.ino:1341]: That would give more flexibility in pin-layout. Although, looking at the suggested pin definition for a Teensy4.0, Serial1 and Serial4 are the only candidates. It would mean re-arranging of the amplifier enable, and bladePowerPin1.
I2C, SPI and I2C are more or lest fixed to certain pins, and also have to make sure your Neopixel aren interfeing wiht any of the 'fixed' pins.

Is making the #define for SERIAL_CONSOLE_PORT  something that you are willing to write and integrate into ProffieOS? (for now, I think the define for Serial4adapter option will work as well)

File: ProffieOS.ino [hunk 3]
```diff
@@ -1646,7 +1666,11 @@ SSD1306 display;
 #ifdef ENABLE_MOTION
 
 #include "motion/motion_util.h"
-#include "motion/mpu6050.h"
+#ifdef USE_TEENSY4
+ #include "motion/mpu6050async.h"
+#else
+ #include "motion/mpu6050.h"
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply ProffieOS.ino:1672]: OK,
 I have looked into making the 'normal' and 'async' reading of the mpu6050 both into the mpu6050,h  file, and have an #if defined(__IMXRT1062__) or #ifdef USE_TEENSY4 switch between the two implementations. The differences are quite big between the two, but would allow to use just 1 include for the mpu6050?
Would that be a better idea?

File: ProffieOS.ino [hunk 4]
```diff
@@ -1773,6 +1797,11 @@ void setup() {
     talkie.Say(talkie_not_found_15, 15);
   }
 #endif // ENABLE_AUDIO && ENABLE_SD
+
+#if defined(USE_TEENSY4) && defined(ENABLE_WS2811)
+  leds.begin(maxLedsPerStrip, displayMemory, drawingMemory, config_, numPins, pinList);
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply ProffieOS.ino:1802]: Yes, I knew when I made it this way that this is one of those 'cringe' parts  :-)
Having the multiple instances of MonopodWSPinBase initializing just one OctoWS2811 instance is where I did not find a good solution. This was the fastest and easiest way for me to get everything working, and leave the OctoWS2811 lib completely untouched.
If you are willing to help, this is were I can you the help of suggestions/examples of how to solve this the correct way.

File: ProffieOS.ino [hunk 5]
```diff
@@ -1799,9 +1828,13 @@ MTPStorage_SerialFlash serialflash_storage(&mtpd);
 
 #include "common/clock_control.h"
 
+uint32_t start_ = millis();
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply ProffieOS.ino:1831]: garbage from experimenting, removed now

File: blades/monopodws.h [hunk 6]
```diff
@@ -86,6 +86,13 @@ class MonopodWS2811 {
   static void show(int pin, uint8_t ones, int bufsize, uint32_t frequency) {
     ones_ = ones;
     
+if(!(init_ & (1 << pin))) {
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply blades/monopodws.h:89]: It is garbage from experimenting, but addresses one of the points I noticed about the behavior of  monopodws;

Each frame-loop the pin is re-initialized, and the complete DMA setup is done over and over again (basically the begin() ) part of OctoWS28111) 

What my version does is keep track of pins calling the MonopodWS2811::show() and check if they already have been initialized, and skip the init part if already doen. It was my hope to get the fps al little bit higher when skipping the init part.

What it does is assign a bit to each pin calling MonopodWS2811::show(), hence:    init_  |=  1 << pin;   (line 228)

if(!(init_ & (1 << pin)))   checks if this pin has been set already and skips the init part if so.

This value is set to 0 on line 367.

Just an idea, but maybe you need to re-init the pin every loop for reasons that I do not undertsand?

File: blades/monopodws_imxrt.h [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+#define BLADES_MONOPODWS_H
+
+#ifdef ENABLE_WS2811
+
+#include <OctoWS2811.h>
+
+// Any group of digital pins may be used
+// Teensy 4.x can use any pin, or arbitrary group of pins!
+// default options {2, 14, 7, 8, 6, 20, 21, 5};
+const int numPins = 3;               //should be auto generated from config file
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply blades/monopodws_imxrt.h:11]: This is also something that I might need a bit more help then just a suggestion :-)
When trying to alter the numPins and pinList during runtime, there were errors, indicating that a const can not be changed, and when making it a non const, the OctoWS2811 lib started complaining it expects a cont for numpins and pinlist.

So my hope was, to have some magic checking the config file pre-compile time and setting the numpins and pinlist maybe even via defines??  

Is it even possible during runtime that the frequency changes? I thought that all WS2811s use 800kHz?

What doe you mean by feeding the blades out in batches? I do not fully understand how that is supposed to work?

File: blades/monopodws_imxrt.h [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  void EndFrame() {
+    
+    while (!IsReadyForEndFrame()); // wait for pixels to be transfered
+    for (int j = 0; j < num_leds_; j++) {
+      Color8 color = color_buffer[j].dither(frame_num_, j); //adjust colors, from common/color.h
+                                                             // color_buffer defined in ws28111_blade.h
+    // For Teensy 4.x, the pixel data is stored in ordinary RGB format. This is
+    // different from Teensy 3.x, where the data was stored as bytes to write directly
+    // to the GPIO output register.      
+    leds.setPixel(j + offset_, color.getByte(BYTEORDER, 0), color.getByte(BYTEORDER, 1), color.getByte(BYTEORDER, 2));
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply blades/monopodws_imxrt.h:110]: I think tyhis should be possible, nbut we might need to trick the OctoWS2811::begin().
It sets:   numbytes = striplen * 3; 
and has a clear comment // TODO: 4 if RGBW

So in order to make sure the numbytes is correct for RGBw and leave the OctoWS2811 lib untouched, might prove tricky?

setting the displayMemory directly from MonopodWSPinBase should be possible I think

File: motion/mpu6050async.h [hunk 9]
```diff
@@ -0,0 +1,201 @@
+#ifndef MOTION_MPU6050_H
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply motion/mpu6050async.h:1]: I think it would be better to have both implementations just in one mpu6050.h file. Using the #ifdef USE_TEENSY4, or #if defined(__IMXRT1062__) to switch between the 2. 
I tried the async implementation on a Teensy3.2 with the #include <i2c_t3.h> I2C, that did not work at all ...
The async implementation requires the regular Wire.h to be the main I2C lib.

I will combine the two implementations into one mpu6050.h and make it a separate pull-request

File: sound/dac.h [hunk 10]
```diff
@@ -1,10 +1,31 @@
 #ifndef SOUND_DAC_H
 #define SOUND_DAC_H
 
-#if defined(__IMXRT1062__)
-void set_audioClock(int nfact, int32_t nmult, uint32_t ndiv,  bool force = false); // sets PLL4
-#endif
+#if defined(__IMXRT1062__) //Teensy 4.0, include, because direct include of imxrt_hw.h did not work??
```
[the remark this answers is not in the document]
Comment (artandtechvof) [reply sound/dac.h:4]: It only worked for me for the version where the #define AUDIO_CONTROL_SGTL5000 was used.
If this is not included, the dac.h failed to find the set_audioClock() implementation (???)

File: blades/monopodws_imxrt.h (hunk 1 above)
Comment (artandtechvof) [reply blades/monopodws_imxrt.h:81]: I have done a lot of experimenting with micros() vs millis(). 

When using micros() I sometimes get strange behavior of the secondary blades. They instantly extend/retract, while the main blade is working as it should. When syncing to full milliseconds rounded upward, I get the most reliable performance of the system.
I also think the performance is good enough even for higher led-counts like the new KR-ledstrips with 264 pixels.
144 pixels -> 4.62ms update (rounded 5ms), 216 fps theoretical vs 200 fps when rounded up to nearest millisec()
264 pixels -> 8.22ms update (rounded 9ms) 121 fps theoretical vs 111 fps when rounded up to nearest millisec()
This is robustness over performance from my point of view, based on empirical evidence...

Comment (profezzorn) [reply blades/monopodws_imxrt.h:81]: The problem might have to do with wraparounds.
This code here for instance, won't when millis() is more than 2^32 - 4
Generally, I find that the only safe way to deal with micros() and millis() is to remember a timestamp, and when you wish to know how long ago that timestamp was, you do micros() - timestamp.  So in this case, that becomes:

  millis() - frame_sync_time_ > min_update_time_

As long as frame_sync_time_ is guaranteed to be in the past (and not TOO long in the past), this always works.

File: sound/dac.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (profezzorn) [reply sound/dac.h:4]: weird, but as long as the audio shield code still works, that's fine

File: ProffieOS.ino (hunk 4 above)
[the remark this answers is not in the document]
Comment (profezzorn) [reply ProffieOS.ino:1798]: Unfortunately, it sounds like leaving OctoWS2811 untouched is not going to be an option that will work.
For a quick hack, this is pretty cool, but there are a bunch of things that won't work properly, which would be relatively easy to implement if we import and modify the OctoWS2811 code.  Like:

1) blade id / blade detect
2) 400/570/1600kHz LEDs (unusual, but they do exist)
3) RGBA
4) actually supporting the way that proffieos defines blades

All of these would be hard or impossible to do while leaving OctoWS2811 the way it is.

File: blades/monopodws_imxrt.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (profezzorn) [reply blades/monopodws_imxrt.h:11]: Basically, I was thinking that we could do something like:

  bool CanUseCurrentBatch() {
     if (!engine.done()) return false;
     if (engine.hasPin(pin_)) return false;  // It already has this pin, feed out the data and start a new batch
     if (frequency_ != engine.frequency()) return false;
     // Check other parameters that must match here.
     return true;
  }

  bool IsReadyForEndFrame() {
      if (!engine.done()) return false;
      if (engine.num_pins() == 0) return true;
      if (CanUseCurrentBatch()) return true;
      // We've determined that we can't use the same batch, feed it out and wait until it's done.
      engine.show();
      return false;
  }

  void EndFrame() {
      while (!IsReadyForEndFrame());
      if (engine.num_pins() == 0) {
          engine.init(frequency_, and other paramters...)
      }
      int p = engine.add_pin(pin_, num_leds_);
      for (int i = 0; i < num_leds_; i++) { engine_.set_data(i, p, color_buffer[i]); }
      if (eigine.num_pins() == numPins) {
        // batch is full, feed out data now
        batch.show();
      }
  }

Basically, batches would be built and send out dynamically, with a maximum of "numPins" strings per batch.
This of course requires making some changes to OctoWS2811 so that pins can be added dynamically and init() can be called over and over again.


Now, pehrhaps a simpler option would be to completely forgoe the multi-string capabilities and do what monopodws does. Since the IMXRT code already have the capability of assembling the DMA data in as it goes, it should be simple to modify it to read directly from the color_buffer, just like the smt32 code does.  Of course, the reason the stm32 code does that is to save memory, which is not really relevant for this processor....

File: ProffieOS.ino (hunk 2 above)
[the remark this answers is not in the document]
Comment (profezzorn) [reply ProffieOS.ino]: SERIAL_CONSOLE_PORT is trivial to add, I can do that part easily.

File: blades/monopodws_imxrt.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (artandtechvof) [reply blades/monopodws_imxrt.h:11]: Ah! Thank you for the explanation, now I understand it better. Only the numPins would still be a constant ste in the (hardware) config file? It would be nice if it was automatically set by the number Neopixel blades in the config file. Just using the NUM_BLADES define would not work, because there might be simple LED or Tri-cree blades in there as well.

Do you have an suggestion of how to derive the number of Neopixel blades from the config file, or during runtime?

Rewriting the OctoWS2811 to make it an "engine" that dynamically updates the pins, is something I can give a go. I really liked the idea of being able to leave the OctoWS2811 untouched. Seems that we need an modified version to be fully ProffieOS compliant.
I also understand better why you are re-initing the blade-pins and all parameters over and over. It makes it more flexible and will update the params if you change blades and have a drifferent blade(-ID) detected?

File: sound/dac.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (artandtechvof) [reply sound/dac.h:4]: Oops... Modified this to only be included if the AUDIO_CONTROL_SGTL5000 has not been defined. Is is now working for the audioshield and a ProffieOS based saber.
(it indeed gave double defined errors when using the audio shield...)

File: blades/monopodws_imxrt.h (hunk 1 above)
Comment (artandtechvof) [reply blades/monopodws_imxrt.h:81]: Yeah, should have done that myself. The way you suggest, the millis() can just roll over but the time difference from millis() - frame_sync_time_   would still be correct because of uint subtraction. 
I will adjust the code to do it like this.

File: blades/monopodws_imxrt.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (profezzorn) [reply blades/monopodws_imxrt.h:11]: > Do you have an suggestion of how to derive the number of Neopixel blades from the config file, or during runtime?

During compilation is difficult, or maybe impossible.
During runtime is easy, it's basically just doing a +=1 in the pin class constructor and a -=1 in the pin class destructor.

**Comment by profezzorn** [comment]
I've been looking at the OctoWS2811 library for other reasons recently.
If i'm not mistaken, it should be possible to run begin() multiple times, and with different arguments, as long as you wait until busy() returns false before calling it again, which might make all of this a lot easier.

**Comment by artandtechvof** [comment]
Thanks for the tip! 
Have to dive into this to see if I understand you correct; Running the begin() with different arguments, would mean making multiple instances (one for each blade) that run independent?
If so, that would certainly help a lot in making this work.

**Comment by profezzorn** [comment]
Most things in this PR has been implemented now.
(Except maybe the MP6050 changes.)
